File: tests/test-minecraft-osp/src/main/java/pize/tests/voxelgame/main/net/packet/CBPacketSpawnEntity.java
package pize.tests.voxelgame.main.net.packet;

import pize.math.util.EulerAngles;
import pize.math.vecmath.vector.Vec3f;
import pize.net.tcp.packet.IPacket;
import pize.net.tcp.packet.PacketHandler;
import pize.util.io.PizeInputStream;
import pize.util.io.PizeOutputStream;
import pize.tests.voxelgame.main.entity.Entity;
import pize.tests.voxelgame.main.entity.EntityType;

import java.io.IOException;
import java.util.UUID;

public class CBPacketSpawnEntity extends IPacket<PacketHandler>{

    public static final int PACKET_ID = 14;

    public CBPacketSpawnEntity(){
        super(PACKET_ID);
    }
    
    public EntityType<?> type;
    public UUID uuid;
    public Vec3f position;
    public EulerAngles rotation;

    public CBPacketSpawnEntity(Entity entity){
        this();
        type = entity.getEntityType();
        uuid = entity.getUUID();
        position = entity.getPosition();
        rotation = entity.getRotation();
    }


    @Override
    protected void write(PizeOutputStream stream) throws IOException{
        stream.writeInt(type.getID());
        stream.writeUTF(uuid.toString());
        stream.writeVec3f(position);
        stream.writeEulerAngles(rotation);
    }

    @Override
    public void read(PizeInputStream stream) throws IOException{
        type = EntityType.fromEntityID(stream.readInt());
        uuid = UUID.fromString(stream.readUTF());
        position = stream.readVec3f();
        rotation = stream.readEulerAngles();
    }
    
}

File: tests/test-minecraft-osp/src/main/kotlin/pize/tests/voxelgame/main/net/packet/CBPacketSpawnEntity.kt
package pize.tests.voxelgame.main.net.packet

import pize.math.util.EulerAngles
import pize.math.vecmath.vector.Vec3f
import pize.net.tcp.packet.IPacket
import pize.net.tcp.packet.PacketHandler
import pize.tests.voxelgame.main.entity.Entity
import pize.tests.voxelgame.main.entity.EntityType
import pize.util.io.PizeInputStream
import pize.util.io.PizeOutputStream
import java.io.IOException
import java.util.*

class CBPacketSpawnEntity() : IPacket<PacketHandler?>(PACKET_ID) {
    var type: EntityType<*>? = null
    var uuid: UUID? = null
    var position: Vec3f? = null
    var rotation: EulerAngles? = null

    constructor(entity: Entity) : this() {
        type = entity.entityType
        uuid = entity.uuid
        position = entity.position
        rotation = entity.rotation
    }

    @Throws(IOException::class)
    override fun write(stream: PizeOutputStream?) {
        stream!!.writeInt(type.getID())
        stream.writeUTF(uuid.toString())
        stream.writeVec3f(position!!)
        stream.writeEulerAngles(rotation!!)
    }

    @Throws(IOException::class)
    override fun read(stream: PizeInputStream?) {
        type = EntityType.Companion.fromEntityID(stream!!.readInt())
        uuid = UUID.fromString(stream.readUTF())
        position = stream.readVec3f()
        rotation = stream.readEulerAngles()
    }

    companion object {
        const val PACKET_ID = 14
    }
}